edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleIdentityLifecycleManager.cs

```diff
@@ -16,27 +16,36 @@ public class ModuleIdentityLifecycleManager : IModuleIdentityLifecycleManager
         readonly IIdentityManager identityManager;
         readonly ModuleIdentityProviderServiceBuilder identityProviderServiceBuilder;
         readonly Uri workloadUri;
+        readonly bool enableExistingIdentityCleanup;
 
         protected virtual bool ShouldAlwaysReturnIdentities => false;
 
-        public ModuleIdentityLifecycleManager(IIdentityManager identityManager, ModuleIdentityProviderServiceBuilder identityProviderServiceBuilder, Uri workloadUri)
+        public ModuleIdentityLifecycleManager(IIdentityManager identityManager, ModuleIdentityProviderServiceBuilder identityProviderServiceBuilder, Uri workloadUri, bool enableExistingIdentityCleanup)
         {
             this.identityManager = Preconditions.CheckNotNull(identityManager, nameof(identityManager));
             this.identityProviderServiceBuilder = Preconditions.CheckNotNull(identityProviderServiceBuilder, nameof(identityProviderServiceBuilder));
             this.workloadUri = Preconditions.CheckNotNull(workloadUri, nameof(workloadUri));
+            this.enableExistingIdentityCleanup = enableExistingIdentityCleanup;
         }
 
         public async Task<IImmutableDictionary<string, IModuleIdentity>> GetModuleIdentitiesAsync(ModuleSet desired, ModuleSet current)
         {
-            Diff diff = desired.Diff(current);
-            if (diff.IsEmpty && !this.ShouldAlwaysReturnIdentities)
-            {
-                return ImmutableDictionary<string, IModuleIdentity>.Empty;
-            }
-
             try
             {
-                IImmutableDictionary<string, IModuleIdentity> moduleIdentities = await this.GetModuleIdentitiesAsync(diff);
+                IImmutableDictionary<string, Identity> identities = (await this.identityManager.GetIdentities()).ToImmutableDictionary(i => i.ModuleId);
+
+                if (this.enableExistingIdentityCleanup)
+                {
+                    await this.RemoveStaleIdentities(desired, current, identities);
+                }
+
+                Diff diff = desired.Diff(current);
+                if (diff.IsEmpty && !this.ShouldAlwaysReturnIdentities)
+                {
+                    return ImmutableDictionary<string, IModuleIdentity>.Empty;
+                }
+
+                IImmutableDictionary<string, IModuleIdentity> moduleIdentities = await this.GetModuleIdentitiesAsync(diff, identities);
```

Comment:
removed identities should be removed from the `identities` dictionary before passing them to `GetModuleIdentitiesAsync`